ringbuf/reader.go

```diff
@@ -203,23 +204,29 @@ func (r *Reader) ReadInto(rec *Record) error {
 
 	for {
 		if !r.haveData {
-			_, err := r.poller.Wait(r.epollEvents[:cap(r.epollEvents)], r.deadline)
-			if err != nil {
-				return err
+			_, err = r.poller.Wait(r.epollEvents[:cap(r.epollEvents)], r.deadline)
+			if errors.Is(err, os.ErrDeadlineExceeded) {
+				r.haveData = !r.ring.isEmpty()
+				if !r.haveData {
+					return err
+				}
+			} else {
+				if err != nil {
+					return err
+				}
+				r.haveData = true
```

Comment:
my original thinking was that something like this would be sufficient:
```suggestion
			_, err := r.poller.Wait(r.epollEvents[:cap(r.epollEvents)], r.deadline)
			if errors.Is(err, os.ErrClosed) || (errors.Is(err, os.ErrDeadlineExceeded) && r.ring.isEmpty()) {
			        return err
			}
```

would that work?

Reply:
Yeah, that is cleaner. two small comments:
1. I changed the err to be in the scope of the function: for the case where we have a timeout and there is available data - I want the err to contain the os.ErrDeadlineExceeded and won't get overriden to nil - such that the caller will know a timeout happened.
2. kept the r.haveData = true after the if statement

Reply:
1. sounds potentially problematic because when you cannot tell if the data is valid based the error?

Reply:
Right, after a second thought I agree.
Updated the new tests accordingly